### PR TITLE
feature: add rule to refactor exception methods

### PR DESCRIPTION
**Pull request by alfredbez**
see: https://github.com/sebastianbergmann/phpunit/issues/3775

### CODE REVIEW

**Comment by alfredbez** [comment]
Is it possible to ensure that we're not breaking expected behavior or should I just remove the fixture?

**Comment by TomasVotruba** [comment]
CI needs fixing

**Comment by alfredbez** [comment]
I know, my question was just about the failing step

**Comment by alfredbez** [comment]
Maybe it makes also sense to include the CI checks to the `complete-check` script in composer, e.g.:
```diff
git --no-pager diff
diff --git a/composer.json b/composer.json
index 826e67b1c..d8494b790 100644
--- a/composer.json
+++ b/composer.json
@@ -213,11 +213,13 @@
     "scripts": {
         "complete-check": [
             "@check-cs",
+            "@check-fixtures",
             "phpunit",
             "@phpstan",
             "@docs"
         ],
         "check-cs": "ecs check bin packages src tests utils compiler --ansi",
+        "check-fixtures": "php ci/check_keep_fixtures.php",
         "fix-cs": [
             "ecs check bin packages src tests utils compiler --fix --ansi",
             "ci/clean_trailing_spaces.sh"
```
Fast feedback is really helpful

**Comment by TomasVotruba** [comment]
> Maybe it makes also sense to include the CI checks to the complete-check script in composer, e.g.:

Yes, why not

File: packages/PHPUnit/tests/Rector/MethodCall/ExplicitPhpErrorApiRector/Fixture/php_combination.php.inc [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+{
+    public function test()
+    {
+        $this->expectException(\PHPUnit\Framework\TestCase\Deprecated::class);
+        $this->expectException(\PHPUnit\Framework\TestCase\Error::class);
+        $this->expectException(\PHPUnit\Framework\TestCase\Notice::class);
+        $this->expectException(\PHPUnit\Framework\TestCase\Warning::class);
+    }
+}
+
```
Comment (TomasVotruba) [inline packages/PHPUnit/tests/Rector/MethodCall/ExplicitPhpErrorApiRector/Fixture/php_combination.php.inc]: This class duplicates the other 4 cases right?
What's the added value?

**Comment by TomasVotruba** [comment]
> Is it possible to ensure that we're not breaking expected behavior or should I just remove the fixture?

I don't know what you mean.

File: packages/PHPUnit/tests/Rector/MethodCall/ExplicitPhpErrorApiRector/Fixture/php_combination.php.inc (hunk 1 above)
Comment (alfredbez) [reply packages/PHPUnit/tests/Rector/MethodCall/ExplicitPhpErrorApiRector/Fixture/php_combination.php.inc]: fixed that

**Comment by alfredbez** [comment]
> > Maybe it makes also sense to include the CI checks to the complete-check script in composer, e.g.:
> 
> Yes, why not

I'll create another PR for that. I rebased this PR again and fixed the fixture, ready for review.

File: packages/PHPUnit/src/Rector/MethodCall/ExplicitPhpErrorApiRector.php [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+     * @param MethodCall|StaticCall $node
+     * @param string $exceptionClass
+     * @param string $explicitMethod
+     */
+    private function replaceExceptionWith(Node $node, $exceptionClass, $explicitMethod): void
+    {
+        if (isset($node->args[0]) && property_exists(
+            $node->args[0]->value,
+            'class'
+        ) && (string) $node->args[0]->value->class === $exceptionClass) {
```
Comment (alfredbez) [inline packages/PHPUnit/src/Rector/MethodCall/ExplicitPhpErrorApiRector.php]: I don't like this part, especially the `property_exists` check, but I'm not sure if there is a better way, to check if `$this->expectException()` has an argument that matches e.g. `PHPUnit\Framework\TestCase\Warning`

Comment (TomasVotruba) [reply packages/PHPUnit/src/Rector/MethodCall/ExplicitPhpErrorApiRector.php]: Let's merge if for now. 
I'll look on this as part of common rule cleanup.

**Comment by TomasVotruba** [comment]
Thank you :+1:

**Pull request merged**